0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,180 @@
+* `han_bitcoin_ethereum_redeem/refund_identity`: Since some identities are exchanged during announcement, the identities are kept in extension tables.
+
+One could optimize here and combine all ledger specific identities into an own table `han_bitcoin_identity` to hold all Bitcoin `refund` and `redeem` identities.
+This would mean though that we would need an additional unique id in those sub-tables, i.e. we can't use the `fk` as `pk`.
+
+* `han_bitcoin/ethereum_ledger_state`: is a new table which we introduced in this database schema.
+It records observed ledger events such as `fund`/`redeem`/`refund`-transaction.
+We extracted the states into `han_state` in order to reduce duplication in each table and not have an extra column for the event-type (`fund`/`redeem`/`refund`).
+This allows us to reuse these states among all `han_*` protocols.
+
+HErc20 and HaLight follow the same principles as described above.
+For completeness we depict `herc20` and `halight` diagrams below.
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/herc20_swap_protocols.puml&fmt=svg[Herc20Protocols]
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/halight_swap_protocols.puml&fmt=svg[HalightProtocols]
+
+=== Swap State
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/swap_states.puml&fmt=svg[SwapStates]
+
+Additional to the specific `alpha` and `beta` ledger state (depicted in <<Protocols>>) a `swap` has two additional states:
+
+. the _communication_ state which depicts the state during the announcement phase, i.e. `sent`, `accepted`, `rejected`
+. the aggregated state of the swap, aggregation of the communication state, ledger states, plus an additional error state, i.e. `in_progress`, `swapped`, `not_swapped`, `internal_failure`
+
+Note, that there is one communication state and one aggregated swap state per swap, whereas the ledger-state is per alpha-/beta-ledger per swap.
+We decided that we want to track multiple communication and aggregated state events, thus we added the joint-tables `swap_communication_state_for_swap` and `swap_state_for_swap` to connect `swap` with the respective state table (`swap_communication_state`, `swap_state`.
+
+Note that all `*_state` tables are static tables similar to the `protocol` table.
+These tables contain fixed values that are only changed if the state-model changes.
+
+== Full DB Model
+
+[[COMITDBSchema]]
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/9e9f70785dd4281c808cc4c7ade8b2d14b133588/assets/0026-database-diagram.puml&fmt=svg[COMITDBSchema]
```

Comment:
What are `sent`, `accepted`, `rejected`?

Reply:
currently Alice `sends` a `swap_request` to Bob who can `accept` or `reject`. Lateron this will be `announce` and `finalized` (and I'm missing something else).

Reply:
You are describing `rfc003` behaviour, but the spike says that the following protocols are in scope:
```
* HAN
* HErc20
* HALight
```

Reply:
Ok clarified, I was being picky on the possible values of the communications state.